traversal/optimiser/Optimiser.java

```diff
@@ -0,0 +1,159 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package com.vaticle.typedb.core.traversal.optimiser;
+
+import com.google.ortools.linearsolver.MPSolver;
+import com.google.ortools.linearsolver.MPSolverParameters;
+import com.google.ortools.linearsolver.MPVariable;
+import com.vaticle.typedb.core.common.exception.TypeDBException;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.ortools.linearsolver.MPSolverParameters.IncrementalityValues.INCREMENTALITY_ON;
+import static com.google.ortools.linearsolver.MPSolverParameters.IntegerParam.INCREMENTALITY;
+import static com.google.ortools.linearsolver.MPSolverParameters.IntegerParam.PRESOLVE;
+import static com.google.ortools.linearsolver.MPSolverParameters.PresolveValues.PRESOLVE_ON;
+import static com.vaticle.typedb.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static com.vaticle.typedb.core.common.iterator.Iterators.iterate;
+
+public class Optimiser {
+
+    private final List<Variable> variables;
+    private final Set<Constraint> constraints;
+    private final Map<Variable, Double> objectiveCoefficients;
+    private ActivationStatus status;
+    private MPSolver solver;
+    private MPSolverParameters parameters;
+
+    public Optimiser() {
+        variables = new ArrayList<>();
+        constraints = new HashSet<>();
+        objectiveCoefficients = new HashMap<>();
+        status = ActivationStatus.INACTIVE;
+    }
+
+    public synchronized ResultStatus solve(long timeLimitMillis) {
+        if (status == ActivationStatus.INACTIVE) activate();
+        solver.setTimeLimit(timeLimitMillis);
+        ResultStatus resultStatus = ResultStatus.of(solver.solve(parameters));
+        variables.forEach(Variable::recordValue);
+        clearInitialisation();
+        return resultStatus;
+    }
+
+
+    public synchronized void deactivate() {
+        solver.delete();
+        parameters.delete();
+        variables.forEach(Variable::deactivate);
+        constraints.forEach(Constraint::deactivate);
+        status = ActivationStatus.INACTIVE;
+    }
+
+    private synchronized void activate() {
+        solver = MPSolver.createSolver("SCIP");
+        solver.objective().setMinimization();
+        parameters = new MPSolverParameters();
+        parameters.setIntegerParam(PRESOLVE, PRESOLVE_ON.swigValue());
+        parameters.setIntegerParam(INCREMENTALITY, INCREMENTALITY_ON.swigValue());
+        variables.forEach(var -> var.activate(solver));
+        constraints.forEach(constraint -> constraint.activate(solver));
+        applyObjective();
+        applyInitialisation();
```

Comment:
when activating the `Optimiser`, apply the variables, constraints, objective, and initialisation to the new `MPSolver` object